train.py

```diff
@@ -0,0 +1,249 @@
+        file_path = f"models/{cfg.file_prefix}-{run_id}-{epoch}.pt"
+        torch.save(
+            {
+                "model_state_dict": module.state_dict(),
+                "cfg": OmegaConf.to_container(cfg),
+            },
+            file_path,
+        )
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+
+        print(f"Epoch {epoch} Validation", flush=True)
+        model.eval()
+        # Evaluate the model on validation set
+        sloss = val_epoch(
+            train_dataloader,
+            model,
+            criterion,
+        )
+        # Log validation and training losses
+        print(sloss)
+        wandb.log({"val/loss": sloss, "train/loss": t_loss})
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+
+    return model, run_id
+
+
+def train_epoch(
+    data_iter: Iterable,
+    model: nn.Module,
+    criterion: Callable,
+    optimizer: torch.optim.Optimizer,
+    scheduler: LambdaLR,
+    train_state: TrainState,
+    pad_idx=2,
+    accum_iter=1,
+) -> tuple[float, TrainState]:
+    start = time.time()
+    total_tokens = 0
+    total_loss = 0
+    tokens = 0
+    n_accum = 0
+    i = 0  # iteration counter
+
+    # create progress bar
+    pbar = tqdm(data_iter)
+    for b in pbar:  # for batch in dataloader
+        batch = Batch(b[0], b[1], pad_idx)
+        encode_decode = model.forward(batch.src, batch.tgt, batch.src_mask, batch.tgt_mask)
+        out = model.generator(encode_decode)
+        loss = criterion(einops.rearrange(out, "b n d -> (b n) d"), einops.rearrange(batch.tgt_y, "b n -> (b n)")) / batch.ntokens
+        loss.backward()
+
+        train_state.step += 1
+        train_state.samples += batch.src.shape[0]
+        train_state.tokens += batch.ntokens
+
+        # Update the model parameters and optimizer gradients every `accum_iter` iterations
+        if i % accum_iter == 0:
+            optimizer.step()
+            optimizer.zero_grad(set_to_none=True)
+            n_accum += 1
+            train_state.accum_step += 1
+        i += 1
+
+        # Update learning rate scheduler
+        scheduler.step()
+
+        # Update loss and token counts
+        total_loss += loss.item()
+        total_tokens += batch.ntokens
+        tokens += batch.ntokens
+
+        # log metrics every 10 steps
+        if i % 10 == 1:
+            lr = optimizer.param_groups[0]["lr"]
+            elapsed = time.time() - start
+            tok_rate = tokens / elapsed
+            pbar.set_description(
+                f"Epoch Step: {i:6d} | Accumulation Step: {n_accum:3d} | Loss: {loss.item():6.2f} | Tokens / Sec {tok_rate:7.1f} | Learning Rate: {lr:6.1e}"
+            )
+
+            # log the loss each to Weights and Biases
+            wandb.log({"train_steps/loss": loss.item()})
+            del batch
+            del b
+    # Return average loss over all tokens and updated train state
+    del loss
```

Comment:
I'm running locally without this, and gpu memory usage indicates some interesting leakage patterns :thinking: 

![image](https://github.com/Nospoko/annotated-transformer/assets/8056825/645a7e7a-784e-4c4e-a548-9cfcf63a646c)

Reply:
I'm guessing that is because I do not del all variables I create in a training loop - I will check that out because I know the way I am doing it now is kinda stupid.

Reply:
Try to focus on my other comments first - I arrived at them when trying to set up GPU debugging, so the idea is that it will be easier to figure this out later :)